File: src/shorthand_string.py
from text_parser import parse_text
#from more_itertools import ( pairwise, split_after, partitions)
import math
import logging
import random
from character import Character


logger = logging.getLogger(__name__)

def split_after(iterable, pred):
    buf = []
    for item in iterable:
        buf.append(item)
        if pred(item):
            yield buf
            buf = []
    if buf:
        yield buf

def pairwise(iterable):
    iterator = iter(iterable)
    a = next(iterator, None)

    for b in iterator:
        yield a, b
        a = b

class Context:
    def __init__(self, x=0, y=0, ref_line=0, right=0, margin=5):
        self.x = x
        self.y = y
        self.ref_line = ref_line
        self.right = right
        self.margin = margin

class Line:
    def __init__(self, chars=None):
        self.chars = chars if chars else []

    @property
    def clauses(self):
        return split_after(self.chars, lambda c: c.is_space)

class Bbox:
    def __init__(self, left, right, top, bottom):
        self.left = left
        self.right = right
        self.top = top
        self.bottom = bottom
        self.width = right - left
        self.height = bottom - top

    def __mul__(self, scale):
        return Bbox(
            self.left * scale,
            self.right * scale,
            self.top * scale,
            self.bottom * scale,
        )

    def __str__(self):
        return f"{self.left} {self.top} {self.width} {self.height}"

class String:
    def __init__(self, text, sh=None):
        self.text = text
        self.sh = sh
        self.chars = self.get_chars(text)
        self.connect()
        self.select_glyphs()
        bbox = self.layout()
        self.svg_template = f"""<svg
            id="svg_root"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
            version="1.1"
            viewBox="{bbox}"
            width="{bbox.width}mm"
            height="{bbox.height}mm"
            preserveAspectRatio="xMinYMin meet"
            data-text=""
            transform="scale(1 1)"
            transform-origin="0 0"
            style="fill: none; stroke: rgb(0, 0, 0); stroke-width: 0.425197; stroke-linecap: round; stroke-linejoin: round; stroke-miterlimit: 4; stroke-opacity: 1; background-color: #f0e68c;" >
            <rect x="{bbox.left}" y="{bbox.top}" width="100%" height="100%" fill="rgb(255 255 255)" />
            {{}}</svg>"""

    #def get_chars(self, text):
    #    chars = []
    #    for word in parse_text(text):
    #        for parts in partitions(word):
    #            subwords = ["".join(part) for part in parts]
    #            if all([subword in self.sh.dictionary for subword in subwords]):
    #                for subword in subwords:
    #                    for char_name in self.sh.dictionary[subword]:
    #                        chars.append(Character(char_name, self.sh))
    #                break
    #        else:
    #            chars.append(Character('Null', self.sh))
    #    return chars
    def get_chars(self, text):
        chars = []
        for word in parse_text(text, self.sh['dictionary']):

            chars.append(Character(word, self.sh))
        return chars

    def connect(self):
        base = None
        idx = 0
        for char in self.chars:
            char.prev = base
            if char.is_marker:
                continue
            elif base:
                base.next = char
            base = char
            char.idx = idx

    def __str__(self):
        return self.create_path_elements()[0]

    def __repr__(self):
        return f"String({'self.text'})"

    def __split(self):
        return [Line(char_list) for char_list in split_after(self.chars, lambda c: c.is_newline)]

    def set_position(self, cntx, char):
        if char.prev is None:
            cntx.x = max(cntx.right - char.left, 0)
            cntx.y = max(cntx.ref_line - char.ascent, 0)

        char.set_pos(cntx.x, cntx.y)
        if char.is_space:
            try:
                cntx.x = cntx.right + char.dx - char.next.left
            except (IndexError, AttributeError):
                cntx.x += char.dx
            try:
                cntx.y = cntx.ref_line + char.dy - char.next.ascent
            except (IndexError, AttributeError):
                cntx.y = cntx.ref_line + char.dy
            cntx.right = cntx.x
        elif char.is_newline:
            try:
                cntx.right = cntx.margin - char.next.left
            except (IndexError, AttributeError):
                pass
            cntx.ref_line += char.dy
            cntx.x = cntx.right
            try:
                cntx.y = cntx.ref_line - char.next.ascent
            except (IndexError, AttributeError):
                cntx.y = cntx.ref_line

        elif char.is_marker:
            try:
                char.set_pos(char.prev.x, char.prev.y)
            except AttributeError:
                pass
        else:
            cntx.x, cntx.y = char.exit
            cntx.right = max(cntx.right, char.x + char.right)

    def correct_left_and_top(self, lines, margin=5):
        for line in lines:
            left_min = min([char.x + char.left for char in line.chars])
            xoffset = margin - left_min if left_min < 0 else 0
            for char in line.chars:
                char.set_pos(char.x + xoffset, char.y)

        top_min = min([char.y + char.top for char in self.chars])
        yoffset = margin - top_min if top_min < 0 else 0
        for char in self.chars:
            char.set_pos(char.x, char.y + yoffset)

    def correct_inter_line(self, lines, margin=5):
        for upper_line, lower_line in pairwise(lines):
            bottom = max([char.y + char.bottom for char in upper_line.chars], default=0)
            top = min([char.y + char.top for char in lower_line.chars], default=bottom)
            if bottom <= top:
                continue
            yoffset = bottom - top + margin
            for char in lower_line.chars:
                char.set_pos(char.x, char.y + yoffset)

    def correct_inter_clause(self, lines, margin=0):
        for line in lines:
            for left_clause, right_clause in pairwise(line.clauses):
                left_clause_right = max([char.x + char.right for char in left_clause], default=0)
                right_clause_left = min([char.x + char.left for char in right_clause], default=left_clause_right)
                if left_clause_right <= right_clause_left:
                    continue
                xoffset = left_clause_right - right_clause_left + margin
                for char in right_clause:
                    char.set_pos(char.x + xoffset, char.y)

    def layout(self, right=5, ref_line=20, margin=5):
        if len(self.chars) == 0:
            return BBox(0, 0, 0, 0)

        cntx = Context(ref_line=ref_line, right=right)

        for char in self.chars:
            self.set_position(cntx, char)

        lines = self.__split()
        self.correct_inter_line(lines, margin=margin)
        self.correct_inter_clause(lines, margin=0)
        self.correct_left_and_top(lines, margin=margin)

        left = math.floor(min([char.x + char.left for char in self.chars])) - margin
        right = math.ceil(max([char.x + char.right for char in self.chars])) + margin
        top = math.floor(min([char.y + char.top for char in self.chars])) - margin
        bottom = math.ceil(max([char.y + char.bottom for char in self.chars])) + margin

        return Bbox(left, right, top, bottom)

    def create_svg(self, to_animate=False, speed_mm_per_s=20, to_generate_keyframe=False, target_s=0):
        marker_buf = []
        char_buf = []
        for char in self.chars:
            if char.is_marker:
                marker_buf.append(char)
            elif char.is_space or char.is_newline:
                char_buf.append(char)
                char_buf.extend(marker_buf)
                marker_buf = []
            else:
                char_buf.append(char)
        char_buf.extend(marker_buf)

        paths = []
        time_s = 0
        for char in char_buf:
            elem, time_s = char.create_path_element(time_s=time_s, to_animate=to_animate, speed_mm_per_s=speed_mm_per_s, to_generate_keyframe=to_generate_keyframe, target_s=target_s)
            #if char.is_space:
            #    time_s += 0.01
            #elif char.is_newline:
            #    time_s += 0.1
            #elif char.is_marker:
            #    time_s += 0.1
            paths.append(elem)

        return self.svg_template.format("".join(paths)), time_s

    def create_key_frames(self, speed_mm_per_s=20, framerate=24):
        total_length = sum((length for c in self.chars for length in c.lengths))
        total_time_s = total_length / speed_mm_per_s
        total_frames = math.ceil(framerate * total_time_s)
        frames = []
        for frame_num in range(total_frames + 4):
            time_s = frame_num * total_time_s / total_frames 
            frames.append(self.create_svg(to_animate=False, speed_mm_per_s=speed_mm_per_s, to_generate_keyframe=True, target_s=time_s)[0])
        return frames

    def select_glyphs(self):
        char_names = [char.name for char in self.chars]
        new_char_names = [char.select_glyph() for char in self.chars]
        count = 0

        while new_char_names != char_names:
            char_names = new_char_names
            new_char_names = [char.select_glyph() for char in self.chars]
            assert count < 10, "Too many iterations in select_glyphs()"
            count += 1

File: src/ShorthandString.js
import { Character } from './Character.js';
import { Context, Line, Bbox, splitAfter, pairwise } from './LayoutComponents.js';
import { parseText } from './TextParser.js'; // パーサーのインポート



export class ShorthandString {
    /**
     * @param {string} text - 入力文字列 (e.g., "がか")
     * @param {Object} sh - データベース全体 (dictionary, trie, character定義を含む)
     */
    constructor(text, sh) {
        this.text = text;
        this.sh = sh;
        
        // 1. 文字列からCharacterの配列を生成
        this.chars = this.getChars(text);
        
        // 2. 双方向リンクの構築 (マーカースキップ構造を内包)
        this.connect();
        
        // 3. グリフの動的選択 (最大10回の収束ループ)
        this.selectGlyphs();
        
        // 4. 2次元衝突回避レイアウトの実行とBboxの決定
        this.bbox = this.layout();

        // Python側の明示的なSVGテンプレート
        this.svgTemplate = (bboxJson) => `<svg
            id="svg_root"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
            version="1.1"
            viewBox="${bboxJson}"
            width="${bboxJson.width}mm"
            height="${bboxJson.height}mm"
            preserveAspectRatio="xMinYMin meet"
            data-text=""
            transform="scale(1 1)"
            transform-origin="0 0"
            style="fill: none; stroke: rgb(0, 0, 0); stroke-width: 0.425197; stroke-linecap: round; stroke-linejoin: round; stroke-miterlimit: 4; stroke-opacity: 1; background-color: #f0e68c;" >
            <rect x="${bboxJson.left}" y="${bboxJson.top}" width="100%" height="100%" fill="rgb(255 255 255)" />
            {paths}</svg>`;
    }

    /**
     * 外部で定義された `parse_text` の結果をもとにCharacterを生成
     * ※お手元の環境の text_parser.js (または同等機能) と連動させてください
     */
    getChars(text) {
        const parsedWords = parseText(text, this.sh['dictionary'], this.sh['trie']);
        console.log(parsedWords);
        return parsedWords.map(word => new Character(word, this.sh));
    }

    /**
     * Pythonの connect() を100%再現
     * 通常文字同士を繋ぎ、マーカー(isMarker)のnext接続はスキップする
     */
    connect() {
        let base = null;
        for (const char of this.chars) {
            char.prev = base;
            if (char.isMarker) {
                continue;
            } else if (base) {
                base.next = char;
            }
            base = char;
        }
    }

    /**
     * Pythonの select_glyphs() を再現
     * 最大10回ループし、すべての文字のグリフ選択状態が変化しなくなるまで収束させる
     */
    selectGlyphs() {
        for (let i = 0; i < 10; i++) {
            let totalChanged = 0;
            for (const char of this.chars) {
                if (char.selectGlyph()) {
                    totalChanged += 1;
                }
            }
            // 変化した文字が0になったら定常状態に達したため終了
            if (totalChanged === 0) {
                return;
            }
        }
        throw new Error("Too many iterations in selectGlyphs()");
    }

    /**
     * Pythonの __split() 相当
     * 改行(isNewline)を基準に Line オブジェクトの配列に分割する
     */
    _split() {
        const lineLists = Array.from(splitAfter(this.chars, c => c.isNewline));
        return lineLists.map(list => new Line(list));
    }

    /**
     * Pythonの set_position(cntx, char) を完全再現
     */
    setPosition(cntx, char) {
        if (char.prev === null) {
            cntx.x = Math.max(cntx.right - char.left, 0);
            cntx.y = Math.max(cntx.refLine - char.ascent, 0);
        }

        char.setPos(cntx.x, cntx.y);

        if (char.isSpace) {
            try {
                cntx.x = cntx.right + char.dx - (char.next ? char.next.left : 0);
            } catch (e) {
                cntx.x += char.dx;
            }
            try {
                cntx.y = cntx.refLine + char.dy - (char.next ? char.next.ascent : 0);
            } catch (e) {
                cntx.y = cntx.refLine + char.dy;
            }
            cntx.right = cntx.x;
        } 
        else if (char.isNewline) {
            try {
                if (char.next) cntx.right = cntx.margin - char.next.left;
            } catch (e) {}
            cntx.refLine += char.dy;
            cntx.x = cntx.right;
            try {
                cntx.y = cntx.refLine - (char.next ? char.next.ascent : 0);
            } catch (e) {
                cntx.y = cntx.refLine;
            }
        } 
        else if (char.isMarker) {
            try {
                if (char.prev) {
                    char.setPos(char.prev.x, char.prev.y);
                }
            } catch (e) {}
        } 
        else {
            const [exitX, exitY] = char.exit;
            cntx.x = exitX;
            cntx.y = exitY;
            cntx.right = Math.max(cntx.right, char.x + char.right);
        }
    }

    correctLeftAndTop(lines, margin = 5) {
        for (const line of lines) {
            const leftMin = Math.min(...line.chars.map(char => char.x + char.left));
            const xoffset = leftMin < 0 ? (margin - leftMin) : 0;
            for (const char of line.chars) {
                char.setPos(char.x + xoffset, char.y);
            }
        }

        const topMin = Math.min(...this.chars.map(char => char.y + char.top));
        const yoffset = topMin < 0 ? (margin - topMin) : 0;
        for (const char of this.chars) {
            char.setPos(char.x, char.y + yoffset);
        }
    }

    correctInterLine(lines, margin = 5) {
        for (const [upperLine, lowerLine] of pairwise(lines)) {
            const bottom = Math.max(...upperLine.chars.map(char => char.y + char.bottom), 0);
            const top = Math.min(...lowerLine.chars.map(char => char.y + char.top), bottom);
            if (bottom <= top) {
                continue;
            }
            const yoffset = bottom - top + margin;
            for (const char of lowerLine.chars) {
                char.setPos(char.x, char.y + yoffset);
            }
        }
    }

    correctInterClause(lines, margin = 0) {
        for (const line of lines) {
            for (const [leftClause, rightClause] of pairwise(line.clauses)) {
                const leftClauseRight = Math.max(...leftClause.map(char => char.x + char.right), 0);
                const rightClauseLeft = Math.min(...rightClause.map(char => char.x + char.left), leftClauseRight);
                if (leftClauseRight <= rightClauseLeft) {
                    continue;
                }
                const xoffset = leftClauseRight - rightClauseLeft + margin;
                for (const char of rightClause) {
                    char.setPos(char.x + xoffset, char.y);
                }
            }
        }
    }

    /**
     * Pythonの layout() を完全再現
     */
    layout(right = 5, refLine = 20, margin = 5) {
        if (this.chars.length === 0) {
            return new Bbox(0, 0, 0, 0);
        }

        const cntx = new Context({ refLine: refLine, right: right, margin: margin });

        for (const char of this.chars) {
            this.setPosition(cntx, char);
        }

        const lines = this._split();
        this.correctInterLine(lines, margin);
        this.correctInterClause(lines, margin);
        this.correctLeftAndTop(lines, margin);

        const left = Math.floor(Math.min(...this.chars.map(char => char.x + char.left))) - margin;
        const rightVal = Math.ceil(Math.max(...this.chars.map(char => char.x + char.right))) + margin;
        const top = Math.floor(Math.min(...this.chars.map(char => char.y + char.top))) - margin;
        const bottom = Math.ceil(Math.max(...this.chars.map(char => char.y + char.bottom))) + margin;

        return new Bbox(left, rightVal, top, bottom);
    }
    /**
     * Pythonの create_svg を100%再現
     * @return {[string, number]} [SVG文字列, 総再生時間(秒)]
     */
    createSvg({ toAnimate = false, speedMmPerS = 20, toGenerateKeyframe = false, targetS = 0 } = {}) {
        const markerBuf = [];
        const charBuf = [];
        
        // 描画順（マーカーを空白や文字の後ろに回す処理）のシミュレート
        for (const char of this.chars) {
            if (char.isMarker) {
                markerBuf.push(char);
            } else if (char.isSpace || char.isNewline) {
                charBuf.push(char);
                charBuf.push(...markerBuf);
                markerBuf.length = 0; // クリア
            } else {
                charBuf.push(char);
            }
        }
        charBuf.push(...markerBuf);

        const paths = [];
        let timeS = 0;
        
        // 各文字のアニメーション要素をタイムラインに沿って生成
        for (const char of charBuf) {
            const [elem, nextTimeS] = char.createPathElement({
                timeS: timeS,
                toAnimate: toAnimate,
                speedMmPerS: speedMmPerS,
                toGenerateKeyframe: toGenerateKeyframe,
                targetS: targetS
            });
            paths.push(elem);
            timeS = nextTimeS;
        }

        const svgContent = this.svgTemplate(this.bbox).replace('{paths}', paths.join(''));
        return [svgContent, timeS];
    }

    /**
     * Pythonの create_key_frames を再現（スライダー等でのシークや動画書き出し用）
     */
    createKeyFrames(speedMmPerS = 20, framerate = 24) {
        // 全パスの総延長を計算
        const totalLength = this.chars.reduce((sum, c) => sum + c.lengths.reduce((lSum, l) => lSum + l, 0), 0);
        const totalTimeS = totalLength / speedMmPerS;
        const totalFrames = Math.ceil(framerate * totalTimeS);
        
        const frames = [];
        for (let frameNum = 0; frameNum < totalFrames + 4; frameNum++) {
            const targetS = (frameNum * totalTimeS) / totalFrames;
            const [svg] = this.createSvg({
                toAnimate: false,
                speedMmPerS: speedMmPerS,
                toGenerateKeyframe: true,
                targetS: targetS
            });
            frames.push(svg);
        }
        return frames;
    }

    toString() {
        return this.createSvg()[0];
    }
}

